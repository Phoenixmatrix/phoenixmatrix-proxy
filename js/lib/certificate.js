import Promise from 'bluebird';
import fs from 'fs';
import { spawn } from 'child_process';

import mkdirp from 'mkdirp';
import helpers from './helpers';
import _  from 'lodash';
import config from './config';

fs = Promise.promisifyAll(fs);
mkdirp = Promise.promisify(mkdirp);

let directory = './certificate/';
let conf = './openssl.cnf';
let caCertPath = directory + 'ca.crt';
let caKeyPath = directory + 'ca.key';
let serialPath = directory + 'serial.seq';

let spawnCommand = function(command, params, input) {
  params = params.slice(0);
  params.unshift(command);
  let process = spawn('openssl', params);
  let output = '';
  let errorOutput = '';

  return new Promise(function(resolve, reject) {
    process.stdout.on('data', function(data) {
      output += data;
    });

    process.stderr.on('data', function(data) {
      errorOutput += data;
    });

    process.on('close', function(code) {
      if(code > 0) {
        console.log(command + ' exited with code ' + code);
        reject(errorOutput);
      } else {
        resolve(output);
      }
    });

    if(input) {
      process.stdin.write(input);
    }
    process.stdin.end();
  });
};

let runningCommand = Promise.resolve();
let spawnCommandSafe = Promise.coroutine(function* (command, params, input) {
  try {
    yield runningCommand;
    return yield (runningCommand = spawnCommand(command, params, input));
  } catch(err) {
    console.log(err);
    runningCommand = Promise.resolve();
  }
});

var domainCertificates = {};
var ca;
var getCA = () => ca;

var createCertificateAuthority = Promise.coroutine(function* () {
  yield spawnCommandSafe('genrsa', ['-out', caKeyPath, '2048']);
  let request = yield spawnCommandSafe('req', ['-new', '-key', caKeyPath, '-config', conf, '-subj', '/CN=phoenixmatrix_do_not_trust/O=do_not_trust_phoenixmatrix/C=US']);
  let certificate = yield spawnCommandSafe('x509', ['-req', '-sha256', '-days', config.certificateExpiration, '-signkey', caKeyPath], request);
  ca = certificate;
  return yield fs.writeFileAsync(caCertPath, certificate);
});

let keyPairPaths = function(domain) {
  return {
    keyPath: `${directory}domains/${domain}.key`,
    certPath: `${directory}domains/${domain}.crt`
  };
};

let readKeyPair = Promise.coroutine(function* (domain) {
  let { keyPath, certPath } = keyPairPaths(domain);
  let [key, certificate] = yield Promise.all([fs.readFileAsync(keyPath), fs.readFileAsync(certPath)]);

  return yield Promise.resolve({ key, certificate });
});

let isCertificateValid = Promise.coroutine(function* (certificate) {
  let info = yield spawnCommandSafe('x509', ['-noout', '-text'], certificate);
  return Date.parse(info.match(/Not After\s?:\s?([^\n]*)\n/)[1]) < Date.now();
});

let getServerCertificate = Promise.coroutine(function* (domain) {
  if(domainCertificates[domain]) {
    return yield Promise.resolve(domainCertificates[domain]);
  }

  let { keyPath, certPath } = keyPairPaths(domain);

  let result = null;

  let exists = yield helpers.exists(keyPath, certPath);
  if(exists) {
    result = yield readKeyPair(domain);
  }

  if(!result || !isCertificateValid(result.certificate)) {
    yield mkdirp(directory + 'domains');
    if(!fs.existsSync(caCertPath)) {
      yield createCertificateAuthority();
    } else {
      ca = yield fs.readFileAsync(caCertPath);
    }

    yield spawnCommandSafe('genrsa', [ '-out', keyPath, '2048']);
    let request = yield spawnCommandSafe('req', ['-new', '-key', keyPath, '-config', conf, '-subj', '/O=do_not_trust_phoenixmatrix/C=US/CN=' + domain]);
    yield spawnCommandSafe('x509', ['-req', '-out', certPath, '-sha256', '-days', config.certificateExpiration, '-CA', caCertPath, '-CAkey', caKeyPath, '-CAcreateserial', '-CAserial', serialPath], request);

    result = yield readKeyPair(domain);
  }

  domainCertificates[domain] = result;
  return yield Promise.resolve(result);
});

exports.createCertificateAuthority = createCertificateAuthority;
exports.getServerCertificate = getServerCertificate;
exports.getCA = getCA;