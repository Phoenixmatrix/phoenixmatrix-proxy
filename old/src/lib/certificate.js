import fsOrig from 'fs';
import crypto from 'crypto';
import {pki, md} from 'node-forge';
import Promise from 'bluebird';
import moment from 'moment';

import mkdirpOrig from 'mkdirp';
import helpers from './helpers';
import config from './config';

const async = Promise.coroutine;

const fs = Promise.promisifyAll(fsOrig);
const mkdirp = Promise.promisify(mkdirpOrig);

const directory = './certificate/';
const caCertPath = directory + 'ca.crt';
const caKeyPath = directory + 'ca.key';

const domainCertificates = {};
let ca;
const getCA = () => ca;

const getSerial = () => crypto.randomBytes(Math.ceil(16 / 2)).toString('hex').slice(0, 16).toUpperCase();
const keys = pki.rsa.generateKeyPair(2048);

const createCertificateAuthority = async(function* () {
  const certificate = pki.createCertificate();
  certificate.publicKey = keys.publicKey;
  certificate.serialNumber = getSerial();
  certificate.validity.notBefore = new Date();
  certificate.validity.notAfter = new Date();
  certificate.validity.notAfter.setFullYear(certificate.validity.notBefore.getFullYear() + 1);
  const attrs = [{
    name: 'commonName',
    value: 'phoenixmatrix_do_not_trust'
  }, {
    name: 'organizationName',
    value: 'do_not_trust_phoenixmatrix'
  }, {
    name: 'countryName',
    value: 'US'
  }];

  certificate.setExtensions([{
    name: 'basicConstraints',
    cA: true
  }]);

  certificate.setSubject(attrs);
  certificate.setIssuer(attrs);
  certificate.sign(keys.privateKey, md.sha256.create());

  ca = pki.certificateToPem(certificate);

  const privateKey = pki.privateKeyToPem(keys.privateKey);
  yield fs.writeFileAsync(caKeyPath, privateKey);
  return yield fs.writeFileAsync(caCertPath, ca);
});

let keyPairPaths = function(domain) {
  return {
    keyPath: `${directory}domains/${domain}.key`,
    certPath: `${directory}domains/${domain}.crt`
  };
};

let readKeyPair = async(function* (domain) {
  let { keyPath, certPath } = keyPairPaths(domain);
  let [key, certificate] = yield Promise.all([fs.readFileAsync(keyPath), fs.readFileAsync(certPath)]);

  return yield Promise.resolve({ key, certificate });
});

let isCertificateValid = async(function* (certPem) {
  const certificate = pki.certificateFromPem(certPem);
  return certificate.validity.notAfter > Date.now();
});

let getServerCertificate = async(function* (domain) {
  if(domainCertificates[domain]) {
    return yield Promise.resolve(domainCertificates[domain]);
  }

  let { keyPath, certPath } = keyPairPaths(domain);

  let result = null;
  let found = yield helpers.exists(keyPath, certPath);
  if(found) {
    result = yield readKeyPair(domain);
  }

  if(!result || !isCertificateValid(result.certificate)) {
    yield mkdirp(directory + 'domains');
    if(!fs.existsSync(caCertPath)) {
      yield createCertificateAuthority();
    } else {
      ca = yield fs.readFileAsync(caCertPath);
    }

    const csr = pki.createCertificationRequest();
    csr.publicKey = keys.publicKey;
    const attrs = [{
      name: 'commonName',
      value: domain
    }, {
      name: 'organizationName',
      value: 'do_not_trust_phoenixmatrix'
    }, {
      name: 'countryName',
      value: 'US'
    }];

    csr.setSubject(attrs);
    csr.sign(keys.privateKey, md.sha256.create());

    const caKeyPem = yield fs.readFileAsync(caKeyPath);
    const caKey = pki.privateKeyFromPem(caKeyPem);
    const caCertPem = yield fs.readFileAsync(caCertPath);
    const caCert = pki.certificateFromPem(caCertPem);

    var certificate = pki.createCertificate();
    certificate.serialNumber = getSerial();
    certificate.validity.notBefore = new Date();

    const expiration = moment(certificate.validity.notBefore);
    expiration.add(config.certificateExpiration, 'days');
    certificate.validity.notAfter = expiration.toDate();

    certificate.setSubject(csr.subject.attributes);
    certificate.setIssuer(caCert.subject.attributes);
    certificate.publicKey = csr.publicKey;
    certificate.sign(caKey, md.sha256.create());

    const serverCertificate = pki.certificateToPem(certificate);
    const serverKey = pki.privateKeyToPem(keys.privateKey);

    yield Promise.all([
      fs.writeFileAsync(certPath, serverCertificate),
      fs.writeFileAsync(keyPath, serverKey)
    ]);

    result = yield Promise.resolve({key: serverKey, certificate: serverCertificate});
  }

  domainCertificates[domain] = result;
  return yield Promise.resolve(result);
});

export default {
  createCertificateAuthority,
  getServerCertificate,
  getCA
}
