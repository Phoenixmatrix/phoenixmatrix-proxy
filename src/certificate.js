// @flow

import fs from 'fs';
import crypto from 'crypto';
import {Observable} from 'rxjs';
import {pki, md} from 'node-forge';

const directory = './certificate/';
const caCertPath = directory + 'ca.crt';
const caKeyPath = directory + 'ca.key';

function getSerial(){
  crypto.randomBytes(Math.ceil(16 / 2)).toString('hex').slice(0, 16).toUpperCase();
}

function getExpirationDate(startDate) {
  const date = new Date();
  date.setFullYear(startDate.getFullYear() + 1);
}

const keyPair$ = Observable.defer(() => {
  return Observable.of(pki.rsa.generateKeyPair(2048));
})
  .publishReplay(1)
  .refCount();

const certificateAuthority$ = keyPair$
  .map((keys) => {
    const startDate = new Date();

    const certificate = pki.createCertificate();
    certificate.publicKey = keys.publicKey;
    certificate.serialNumber = getSerial();
    certificate.validity.notBefore = startDate;
    certificate.validity.notAfter = getExpirationDate(startDate);
    certificate.setExtensions([{
      name: 'basicConstraints',
      cA: true
    }]);

    const attributes = [
      {
        name: 'commonName',
        value: 'phoenixmatrix_do_not_trust'
      },
      {
        name: 'organizationName',
        value: 'do_not_trust_phoenixmatrix'
      },
      {
        name: 'countryName',
        value: 'US'
      }
    ];

    certificate.setSubjects(attributes);
    certificate.setIssuer(attributes);
    certificate.sign(keys.privateKey, md.sha256.create());

    const ca = pki.certificateToPem(certificate);
    const privateKey = pki.privateKeyToPem(keys.privateKey);

    return {privateKey, ca};
  }).mergeMap(({privateKey, ca}) => {
    const writePrivateKey$ = Observable.bindNodeCallback(
      fs.writeFile(caKeyPath, privateKey)
    );

    const writeCA$ = Observable.bindNodeCallback(
      fs.writeFile(caCertPath, ca)
    );

    return Observable
      .forkJoin(writePrivateKey$, writeCA$)
      .map(() => ({privateKey, ca}));
  });


export {
  createCA
};
