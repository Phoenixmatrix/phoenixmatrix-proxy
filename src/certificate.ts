import * as crypto from 'crypto';
import {pki, md} from 'node-forge';

type KeyPair = {privateKey: string, publicKey: string};
type CertificateAuthorityPair = {privateKey: string, certificate: string};

function getSerial(){
  crypto.randomBytes(Math.ceil(16 / 2)).toString('hex').slice(0, 16).toUpperCase();
}

function getExpirationDate(startDate: Date) {
  const date = new Date(startDate.getTime);
  date.setFullYear(startDate.getFullYear() + 1);
}

const generateKeyPair = (): string => pki.rsa.generateKeyPair(2048);

function createCertificateAuthority(keyPair: KeyPair): CertificateAuthorityPair {
    const startDate = new Date();
    const certificate = pki.createCertificate();
    certificate.publicKey = keyPair.publicKey;
    certificate.serialNumber = getSerial();
    certificate.validity.notBefore = startDate;
    certificate.validity.notAfter = getExpirationDate(startDate);
    certificate.setExtensions([{
      name: 'basicConstraints',
      cA: true
    }]);

    const attributes = [
      {
        name: 'commonName',
        value: 'phoenixmatrix_do_not_trust'
      },
      {
        name: 'organizationName',
        value: 'do_not_trust_phoenixmatrix'
      },
      {
        name: 'countryName',
        value: 'US'
      }
    ];

    certificate.setSubjects(attributes);
    certificate.setIssuer(attributes);
    certificate.sign(keyPair.privateKey, md.sha256.create());

    const ca = pki.certificateToPem(certificate);
    const privateKey = pki.privateKeyToPem(keyPair.privateKey);

    return {privateKey, certificate: ca};
}

export {
  generateKeyPair,
  createCertificateAuthority
};
