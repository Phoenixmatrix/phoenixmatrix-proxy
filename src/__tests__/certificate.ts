import {generateKeyPair, createCertificateAuthority} from '../certificate';

describe('generateKeyPair', () => {
  it('creates a key pair', () => {
    const keyPair = generateKeyPair();
    expect(keyPair).toMatchSnapshot();
  });
});

describe('createCertificateAuthority', () => {
  it('it generates a private key and certificate pair', () => {
    const keyPair = generateKeyPair();

    const certificateAuthority = createCertificateAuthority(keyPair);
  });
});