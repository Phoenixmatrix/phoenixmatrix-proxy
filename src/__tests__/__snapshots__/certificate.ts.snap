exports[`generateKeyPair creates a key pair 1`] = `
Object {
  "privateKey": Object {},
  "publicKey": Object {},
}
`;
